pallets/pips/src/lib.rs

```diff
@@ -1581,55 +1598,98 @@ impl<T: Config> Pallet<T> {
+        state: ProposalState,
+        prune: bool,
+    ) -> Result<u32, DispatchError> {
+        let n_votes = Self::refund_proposal(did, pip_id)?;
         Self::decrement_count_if_active(state);
         if prune {
-            ProposalResult::<T>::remove(id);
-            #[allow(deprecated)]
-            ProposalVotes::<T>::remove_prefix(id, None);
-            ProposalMetadata::<T>::remove(id);
-            if let Some(Proposer::Committee(_)) = Self::proposals(id).map(|p| p.proposer) {
-                CommitteePips::<T>::mutate(|list| list.retain(|&i| i != id));
+            ProposalResult::<T>::remove(pip_id);
+            Self::vote_pruning(pip_id)?;
+            ProposalMetadata::<T>::remove(pip_id);
+            if let Some(Proposer::Committee(_)) = Proposals::<T>::get(pip_id).map(|p| p.proposer) {
+                CommitteePips::<T>::mutate(|list| list.retain(|&i| i != pip_id));
             }
-            Proposals::<T>::remove(id);
-            PipSkipCount::<T>::remove(id);
-            ProposalStates::<T>::remove(id);
+            Proposals::<T>::remove(pip_id);
+            PipSkipCount::<T>::remove(pip_id);
+            ProposalStates::<T>::remove(pip_id);
         }
-        Self::deposit_event(Event::PipClosed(did, id, prune));
-        Ok(())
+        Self::deposit_event(Event::PipClosed(did, pip_id, prune));
+        Ok(n_votes)
     }
 
-    /// Refunds any tokens used to vote or bond a proposal.
-    ///
-    /// This operation is idempotent wrt. chain state,
-    /// i.e., once run, refunding again will refund nothing.
-    fn refund_proposal(did: IdentityId, id: PipId) -> DispatchResult {
-        let mut total_refund = 0;
-        for (_, deposit) in Deposits::<T>::drain_prefix(id) {
-            Self::reduce_lock(&deposit.owner, deposit.amount)?;
-            total_refund = total_refund.saturating_add(deposit.amount);
+    /// Refunds up to [`MaxDepositsRefunded`] deposits for the given `pip_id`.
+    /// Returns the number of refunds processed (it should be the same as the number of votes or equal to MaxRefundsAndVotesPruned).
+    fn refund_proposal(did: IdentityId, pip_id: PipId) -> Result<u32, DispatchError> {
+        let max_refunds = T::MaxRefundsAndVotesPruned::get() as usize;
+
+        let mut n_refunds = 0;
+        let mut refunded_amount = 0;
+        for (_, deposit_info) in Deposits::<T>::drain_prefix(pip_id).take(max_refunds) {
+            n_refunds = n_refunds.saturating_add(1);
+            Self::reduce_lock(&deposit_info.owner, deposit_info.amount)?;
+            refunded_amount = refunded_amount.saturating_add(deposit_info.amount);
+        }
+
+        // Checks if there are more deposits to refund
+        if Deposits::<T>::iter_prefix(pip_id).next().is_some() {
+            PendingRefunds::<T>::try_mutate(|v| v.try_push(pip_id))
+                .map_err(|_| Error::<T>::PendingQueueIsFull)?;
+        }
+
+        Self::deposit_event(Event::ProposalRefund(did, pip_id, refunded_amount));
```

Comment:
I think we should only be doing the refunding (and vote pruning) in the `on_idle` handler.  Here we should just add the pip id to a queue (a storage map, not a vec) of pips to refund & prune votes.  This would also simplify the benchmarks for extrinsics that execute/reject/prune pips, since they wouldn't need the `MaxRefundsAndVotesPruned` weight.  In the `on_idle` handler, we would just copy X pips from the pending queue and start refunding/draining the `Deposits` and `ProposalVotes` storages for that pip.  When the draining is finished, then we remove that pip from the pending queue.